apps/src/blockly/addons/cdoXml.ts

```diff
@@ -455,19 +455,50 @@ export function getBlockElements(xml: Element) {
   return Array.from(xml.querySelectorAll('xml > block'));
 }
 
+/**
+ * Removes all top-level invisible blocks from a Blockly xml element. "Top-level" blocks
+ * in this case also includes blocks that nested directly within a category element.
+ * Used to (unsupported) invisible blocks from a toolbox definition
+ */
 export function removeInvisibleBlocks(xml: Element) {
-  // Get all top-level blocks
-  const topLevelBlocks = xml.getElementsByTagName('block');
+  const categories = xml.getElementsByTagName('category');
+  const blocks = xml.getElementsByTagName('block');
+
   // Convert HTMLCollection to an array to iterate safely
-  const blocksArray = Array.from(topLevelBlocks);
+  const blocksArray = Array.from(blocks);
+  const categoriesArray = Array.from(categories);
 
   blocksArray.forEach(block => {
     if (
-      block.parentElement === xml &&
+      (block.parentElement === xml ||
+        categoriesArray.includes(block.parentElement as Element)) &&
       block.getAttribute('uservisible') === 'false'
     ) {
       block.remove();
     }
   });
   return xml;
 }
+
+/**
+ * Removes statically-defined procedure call blocks from the auto-populated
+ * Functions toolbox category. Call blocks are automatically supplied by the
+ * flyout category callback.
+ */
+export function removeStaticCallBlocks(xml: Element) {
+  // Find the auto-populated Functions category, if it exists.
+  const procedureCategory = Array.from(
+    xml.getElementsByTagName('category')
+  ).find(category => category.getAttribute('custom') === 'PROCEDURE');
+
+  if (procedureCategory) {
+    // Find any procedure call blocks defined within the XML for this category
+    const procedureCallBlocks = Array.from(
+      procedureCategory.getElementsByTagName('block')
+    ).filter(block => block.getAttribute('type') === BLOCK_TYPES.procedureCall);
+
+    // Remove each of the filtered blocks
+    procedureCallBlocks.forEach(block => block.remove());
```

Comment:
Are there ever cases where we want the statically defined blocks? Thinking of levels that start with some functions pre-defined, are those still automatically created?

Reply:
Those are still automatically created as long as we're using the auto-populated category (`custom = 'PROCEDURE'`). The most correct way to provide pre-defined functions is to add the definitions to the start blocks and then just let the category callback create the call blocks for the toolbox automatically.

In levels where we don't want students to create or edit pre-defined functions (e.g. Minecraft Hero HoC), we can use statically defined call blocks, but these are placed in a static category (no auto-population) instead. These levels are unaffected because the category element won't have the `custom` attribute.